src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -461,6 +461,10 @@ oop G1ParScanThreadState::do_copy_to_survivor_space(G1HeapRegionAttr const regio
   Klass* klass = old->klass();
   const size_t word_sz = old->size_given_klass(klass);
 
+  if (region_attr.is_pinned() && klass->is_typeArray_klass()) {
+    return handle_evacuation_failure_par(old, old_mark, word_sz, true /* cause_pinned */);
+  }
```

Comment:
I think it would make sense to add a comment here that this is an optimization, that we know that only type array can be pinned so we only care if the region is pinned or not for type arrays.